torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -435,11 +439,61 @@ std::vector<at::Tensor> ProcessGroupGloo::AsyncWork::result() {
       "Work needs to be completed before calling result(). "
       "Should call wait() before result().");
   TORCH_CHECK(
-      outputTensors_.size() <= 1, "work result does not support list of lists.");
+      outputTensors_.size() <= 1,
+      "work result does not support list of lists, use .getFuture() and value()");
   return outputTensors_.size() == 0 ? std::vector<at::Tensor>()
                                     : outputTensors_.at(0);
 }
 
+c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupGloo::AsyncWork::
+    getFuture() {
+  return future_;
+}
+
+namespace {
+c10::intrusive_ptr<c10::ivalue::Future> CreateFutureForOutput(
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() > 1) {
+    return c10::make_intrusive<c10::ivalue::Future>(
+        c10::ListType::create(c10::ListType::create(c10::TensorType::get())));
+  }
+  return c10::make_intrusive<c10::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()));
+}
+
+void ReturnFutureWithOutput(
+    c10::intrusive_ptr<c10::ivalue::Future> future,
+    const std::vector<std::vector<at::Tensor>>& outputTensors) {
+  if (outputTensors.size() == 0) {
+    future->markCompleted(c10::IValue(std::vector<at::Tensor>()));
+    return;
+  }
+  if (outputTensors.size() > 1) {
+    future->markCompleted(c10::IValue(outputTensors));
+    return;
+  }
+  future->markCompleted(c10::IValue(outputTensors[0]));
+}
+} // namespace
+
+ProcessGroupGloo::AsyncWork::AsyncWork(
+    std::vector<std::vector<at::Tensor>> outputTensors,
+    const char* profilingTitle,
+    const c10::optional<std::vector<at::Tensor>>& inputTensors)
+    : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
+      outputTensors_(std::move(outputTensors)),
+      future_(CreateFutureForOutput(outputTensors)) {}
+
+void ProcessGroupGloo::AsyncWork::finishWorkGlooError(std::exception_ptr eptr) {
+  future_->setError(eptr);
+  finish(eptr);
```

Comment:
similarly, since `ivalue::Future` has similar concepts and states for `completed_` / `exception_`, shall we just use that instead? cc @lw

Reply:
we exposed completed_  and exception_ as protected members of ProcessGroup::Work, so potentially they can be referred to by 3rd parties. Let's make deprecation announcement for 1.10 and then it'd be easier to make better Work/Future merge which can also include CUDA synchronization.

Reply:
I don't get it: how can third-parties access the `completed_` and `exception_` fields? If they are protected it means they can only be accessed by subclasses, but I don't see the point of users subclassing Work: even if they did, ProcessGroupGloo would still return its own classes, and not the user ones, so what does one gain? Moreover, the trailing underscore in the fields names is C++'s "convention" for private fields. All in all, I don't think we should worry about users depending on it.

I agree with Shen that now, effectively, we're doing the same thing twice within Work: we have two ways to mark completion (Work's own one, and the one within Future), etc. One way to go about it is to redefine Work so that it just "forwards" each of its methods to the Future it holds (e.g., Work::wait() basically just calls this->future_->wait()).

Reply:
@lw 

> I don't get it: how can third-parties access the `completed_` and `exception_` fields? If they are protected it means they can only be accessed by subclasses, but I don't see the point of users subclassing Work: even if they did, ProcessGroupGloo would still return its own classes, and not the user ones, so what does one gain?

MPI/GLOO subclass Work, same way 3rd party implementers can also subclass Work and use `completed_` and `exception_`?

ProcessGroupGloo ProcessGroupMPI Work subclasses use (and 3rd party can) PG::Work's `completed_` and `exception_`, so if we remove them 3rd party Work classes will break.

> Moreover, the trailing underscore in the fields names is C++'s "convention" for private fields. 

I thought trailing underscore just means it's a member, not necessarily private?

> All in all, I don't think we should worry about users depending on it.

I think you are right, but it's good to be cautious.

> I agree with Shen that now, effectively, we're doing the same thing twice within Work: we have two ways to mark completion (Work's own one, and the one within Future), etc. One way to go about it is to redefine Work so that it just "forwards" each of its methods to the Future it holds (e.g., Work::wait() basically just calls this->future_->wait()).

the plan is to remove fields from PG::Work and replace it a future. Note that right now, MPI and GLOO subclasses hold a future.

Reply:
Right, I hadn't understood the `completed_` and `exception_` fields came from the root "abstract" base class Work (which, well, isn't actually abstract it seems). Indeed, people who created their own ProcessGroup implementations (as extensions) would be broken by any change there. But I think we eventually will need to do it, right?